server/src/quotas.rs

```diff
@@ -0,0 +1,373 @@
+use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
+
+use failure::Fail;
+use r2d2::Pool;
+
+use crate::actors::project::{Quota, RedisQuota, RejectAllQuota, RetryAfter};
+
+/// The ``grace`` period allows accomodating for clock drift in TTL
+/// calculation since the clock on the Redis instance used to store quota
+/// metrics may not be in sync with the computer running this code.
+static GRACE: u64 = 60;
+
+#[derive(Debug, Clone, Copy)]
+struct UnixTimestamp(u64);
+
+#[derive(Clone)]
+pub enum RedisPool {
+    Cluster(Pool<redis::cluster::ClusterClient>),
+    Single(Pool<redis::Client>),
+}
+
+impl RedisPool {
+    pub fn cluster(servers: Vec<&str>) -> Result<Self, QuotasError> {
+        Ok(RedisPool::Cluster(
+            Pool::builder()
+                .max_size(24)
+                .build(redis::cluster::ClusterClient::open(servers).map_err(QuotasError::Redis)?)
+                .map_err(QuotasError::RedisPool)?,
+        ))
+    }
+
+    pub fn single(server: &str) -> Result<Self, QuotasError> {
+        Ok(RedisPool::Single(
+            Pool::builder()
+                .max_size(24)
+                .build(redis::Client::open(server).map_err(QuotasError::Redis)?)
+                .map_err(QuotasError::RedisPool)?,
+        ))
+    }
+}
+
+#[derive(Debug, Fail)]
+pub enum QuotasError {
+    #[fail(display = "failed to connect to redis")]
+    RedisPool(#[cause] r2d2::Error),
+
+    #[fail(display = "failed to talk to redis")]
+    Redis(#[cause] redis::RedisError),
+}
+
+lazy_static::lazy_static! {
+    static ref IS_RATE_LIMITED: redis::Script = redis::Script::new(include_str!("is_rate_limited.lua"));
```

Comment:
In tests there are poisoning errors, I believe around this script. If we make this a property on some sort of service object, this might also no longer happen. 

Generally, I think we should try to avoid `lazy_static`. When working on Symbolicator, it showed that lazy statics also make testing much harder.